src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ModuleInitializerAttributeShouldNotBeUsedInLibraries.cs

```diff
@@ -0,0 +1,80 @@
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.ModuleInitializerAttributeShouldNotBeUsedInLibrariesMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.ModuleInitializerAttributeShouldNotBeUsedInLibrariesDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                    s_localizableTitle,
+                                                                    s_localizableMessage,
+                                                                    DiagnosticCategory.Usage,
+                                                                    RuleLevel.BuildWarning,
+                                                                    s_localizableDescription,
+                                                                    isPortedFxCopRule: true,
+                                                                    isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                INamedTypeSymbol? moduleInitializerAttributeType = compilationContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeCompilerServicesModuleInitializerAttribute);
+
+                if (moduleInitializerAttributeType is null) return;
+
+                compilationContext.RegisterSymbolAction(context =>
+                {
+                    if (context.Symbol is IMethodSymbol method)
+                    {
+                        if (method.IsPrivate() ||
+                            method.Parameters.Length > 0 ||
+                            method.IsGenericMethod ||
+                            !method.IsStatic ||
+                            !method.ReturnsVoid
+                        )
+                        {
+                            return;
+                        }
+
+                        AttributeData? initializerAttribute = context.Symbol.GetAttributes(moduleInitializerAttributeType).FirstOrDefault();
+
+                        if (initializerAttribute is not null)
+                        {
+                            context.ReportDiagnostic(initializerAttribute.ApplicationSyntaxReference.GetSyntax().CreateDiagnostic(Rule));
```

Comment:
nit:

```suggestion
                            context.ReportDiagnostic(initializerAttribute.ApplicationSyntaxReference.CreateDiagnostic(Rule));
```

Reply:
`SyntaxReference` doesn't have `CreateDiagnostic` on it and I couldn't find an extension method available for this shortcut. But I am going to edit this to pass the cancellation token into `GetSyntax` at least.

Reply:
@jeffhandley Oh, sorry. It's defined in Roslyn.Diagnostics.Analyzers project, but not for NetAnalyzers

https://github.com/dotnet/roslyn-analyzers/blob/main/src/Roslyn.Diagnostics.Analyzers/Core/DiagnosticExtensions.cs

Reply:
Ah, gotcha. OK if I leave it as-is then, or would you prefer I make that extension method available?

Reply:
I think it's okay to leave it.